src/test/run-make-fulldeps/rustdoc-json/compare.py

```diff
@@ -0,0 +1,105 @@
+# relative ordering to check them against eachother by looking them up in their respective blob's
+# `index` or `paths` mappings. To add a new test run `rustdoc --output-format json -o . yourtest.rs`
+# and then create `yourtest.expected` by stripping unnecessary details from `yourtest.json`.
+
+import sys
+import json
+import types
+
+# Used instead of the string ids when used as references.
+# Not used as keys in `index` or `paths`
+class ID(str):
+    pass
+
+
+class SubsetException(Exception):
+    def __init__(self, msg, trace):
+        self.msg = msg
+        self.trace = msg
+        super().__init__("{}: {}".format(trace, msg))
+
+
+def check_subset(expected_main, actual_main):
+    expected_index = expected_main["index"]
+    expected_paths = expected_main["paths"]
+    actual_index = actual_main["index"]
+    actual_paths = actual_main["paths"]
+    already_checked = set()
+
+    def _check_subset(expected, actual, trace):
+        expected_type = type(expected)
+        actual_type = type(actual)
+        if expected_type is not actual_type:
+            raise SubsetException(
+                "expected type `{}`, got `{}`".format(expected_type, actual_type), trace
+            )
+        if expected_type in (str, int, bool) and expected != actual:
+            raise SubsetException("expected `{}`, got: `{}`".format(expected, actual), trace)
+        if expected_type is dict:
+            for key in expected:
+                if key not in actual:
+                    raise SubsetException("Key `{}` not found in output".format(key), trace)
+                new_trace = trace.copy()
+                new_trace.append(key)
+                _check_subset(expected[key], actual[key], new_trace)
+        elif expected_type is list:
+            expected_elements = len(expected)
+            actual_elements = len(actual)
+            if expected_elements != actual_elements:
+                raise SubsetException(
+                    "Found {} items, expected {}".format(expected_elements, actual_elements), trace
+                )
+            for expected, actual in zip(expected, actual):
+                new_trace = trace.copy()
+                new_trace.append(expected)
+                _check_subset(expected, actual, new_trace)
+        elif expected_type is ID and expected not in already_checked:
+            already_checked.add(expected)
+            _check_subset(expected_index.get(expected, {}), actual_index.get(actual, {}), trace)
+            _check_subset(expected_paths.get(expected, {}), actual_paths.get(actual, {}), trace)
+
+    _check_subset(expected_main, actual_main, [])
+
+
+def rustdoc_object_hook(obj):
+    # No need to convert paths, index and external_crates keys to ids, since
+    # they are the target of resolution, and never a source itself.
+    if "id" in obj and obj["id"]:
+        obj["id"] = ID(id)
+    if "root" in obj:
+        obj["root"] = ID(id)
```

Comment:
Not sure what this is trying to do, but it's giving an error that `id` doesn't exist.
```
------------------------------------------
stderr:
------------------------------------------
Traceback (most recent call last):
  File "compare.py", line 105, in <module>
    main(sys.argv[1], sys.argv[2])
  File "compare.py", line 95, in main
    expected_main = json.load(expected_file, object_hook=rustdoc_object_hook)
  File "/usr/lib/python2.7/json/__init__.py", line 291, in load
    **kw)
  File "/usr/lib/python2.7/json/__init__.py", line 352, in loads
    return cls(encoding=encoding, **kw).decode(s)
  File "/usr/lib/python2.7/json/decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python2.7/json/decoder.py", line 380, in raw_decode
    obj, end = self.scan_once(s, idx)
  File "compare.py", line 74, in rustdoc_object_hook
    obj["root"] = ID(id)
UnboundLocalError: local variable 'id' referenced before assignment
```

Reply:
Whoops I committed that too early, I need to rebase and I'll push the working one.